SU2_CFD/include/solver_structure.inl

```diff
@@ -2438,16 +2435,15 @@ inline void CTurbSolver::SetInlet_TurbVar(unsigned short val_marker, unsigned lo
 }
 
 inline void CTurbSASolver::SetFreeStream_Solution(CConfig *config) {
-  for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++)
-    node[iPoint]->SetSolution(0, nu_tilde_Inf);
+  for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++) node->SetSolution(iPoint, 0, nu_tilde_Inf);
 }
 
 inline su2double CTurbSASolver::GetNuTilde_Inf(void) { return nu_tilde_Inf; }
 
 inline void CTurbSSTSolver::SetFreeStream_Solution(CConfig *config){
   for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++){
-    node[iPoint]->SetSolution(0, kine_Inf);
-    node[iPoint]->SetSolution(1, omega_Inf);
+    node->SetSolution(iPoint, 0, kine_Inf);
+    node->SetSolution(iPoint, 1, omega_Inf);
```

Comment:
Which means that in a lot of places we have this kind of change.
In practice some methods of CVariable (e.g. moving the solution between time levels) could now operate on all points which could simplify some areas of the code and improve performance, but for now I think it is better to make the changes as simple as possible to minimize conflicts with other developments.

I tested this on a simple 2D case with 130k cells (Roe + SST) and it was about 30% faster. Not world shattering, I know, but hey it only took 3 days of search and replace xD

Reply:
👍 nice job. This is roughly what we saw for the optimizations performed for the C&F paper. We can check w a profiler to see how much you have improved the cache efficiency.. at some point we know we will be memory bandwidth-bound in the FVM solver, but I think we still have some more room to keep improving things until we hit that point.

Btw, the methods you mention that could operate on all points within CVariable could be good candidates to prototype some SIMD vectorization, since the implementation could be hidden inside CVariable

Reply:
Thanks, I played around with perf a bit already. I am trying to put together a representative case to get some numbers and help decide where to strike next.
Most of the methods I moved are array copies, clang and icc vectorize those, gcc is a bit more paranoid, no luck without pragmas.

Reply:
I have a few ideas if you are looking for some more targets...

You have already made memory access more efficient, so another option to improve performance is to increase the arithmetic intensity per access. For the FVM solver, our algorithms leave us with low flop/byte by nature, but there are still a few things we can try. With relatively minor adjustments, I think we can fuse some kernels to give us more work per edge/point loop. Any edge/point loops that can be processed concurrently could be merged, again as long as it doesn't hurt our flexibility/readability too badly. Goal would be to minimize the total number of these loops per iteration of the solver while stacking up as much compute as possible inside them.

For example, the Upwind_Residual and Viscous_Residual routines could be fused such that one edge loop takes care of both calls to the inviscid and viscous ComputeResidual() routines, since all of the data for both should be readily available. This could be done for other edge/point loops during the preprocessing (limiters, updating primitives, gradients), as long as we are careful with correctness (the order and MPI are important) and don't limit our choice of algorithms.

Also: it would be good to verify that the compiler is able to correctly unroll any loops for nDim and nVar down in the compute intensive kernels.

Reply:
I am glad you bring up fusing Upwind and Viscous residual as that is definitely where the next "easy" performance gain is hiding, and we would need that to make good use of SIMD, it's all about the flop/byte like you say.
It may be relatively simple to do this and another benefit would be lower discrete adjoint memory footprint (half the flux Jacobians to pre-accumulate).

I never thought about fusing gradients and limiters though, I guess that would require looping by cells instead of by edges. One problem I see is that sometimes we want gradients and limiters, but to compute turbulence source terms we only want gradients (and there we are already a bit inefficient as we compute the gradient of all primitives and just the velocities would do), in the interest of readability/maintenance I do not know if we want to end up two similar routines.

Loop unrolling can go terribly wrong (i.e. code bloat that can increase instruction cache misses) for small sizes if the compiler does not have some solid information about loop size for [example](https://gcc.godbolt.org/z/eM2I13) compilers do very different things with those two loops, the second one is unrolled nicely and icc even partly vectorizes, for the generic loop 90% of the instructions would be cold all the time.